File: Grant_engine-2/app/db/queries.py
"""All MongoDB read functions used by the Streamlit UI.

Streamlit reads directly from MongoDB — no API layer needed.
"""
from __future__ import annotations

import os
from datetime import datetime, timezone
from typing import Any, Dict, List, Optional

from motor.motor_asyncio import AsyncIOMotorClient
from pymongo import MongoClient

# Streamlit uses sync PyMongo (Streamlit isn't async)
_client: MongoClient | None = None


def _db():
    global _client
    if _client is None:
        _client = MongoClient(os.environ.get("MONGODB_URI", "mongodb://localhost:27017"))
    return _client["altcarbon_grants"]


# ── Raw scrape preview ─────────────────────────────────────────────────────────

def get_raw_grants_preview(limit: int = 20) -> List[Dict]:
    """Return recently scraped raw grants (before analyst scoring)."""
    db = _db()
    docs = list(
        db["grants_raw"].find(
            {},
            {"title": 1, "url": 1, "funder": 1, "source": 1,
             "themes_detected": 1, "scraped_at": 1, "raw_content": 1, "processed": 1},
            sort=[("scraped_at", -1)],
        ).limit(limit)
    )
    for d in docs:
        d["_id"] = str(d["_id"])
        # Truncate raw_content to a short snippet for display
        content = d.get("raw_content") or ""
        d["snippet"] = content[:300].strip()
        d.pop("raw_content", None)
    return docs


def get_raw_stats() -> Dict:
    db = _db()
    total = db["grants_raw"].count_documents({})
    unprocessed = db["grants_raw"].count_documents({"processed": False})
    scored = db["grants_scored"].count_documents({})
    return {"total_raw": total, "unprocessed": unprocessed, "scored": scored}


# ── Dashboard ──────────────────────────────────────────────────────────────────

def get_tracker_grants(
    search: str = "",
    theme_filter: str = "",
    grant_type_filter: str = "",
    status_filter: str = "",
    min_funding: Optional[int] = None,
    max_funding: Optional[int] = None,
) -> List[Dict]:
    """Return all actionable grants (triage + pursue + watch) sorted by score,
    with all fields needed for the Grant Tracker table."""
    db = _db()
    query: Dict[str, Any] = {
        "status": {"$in": ["triage", "pursue", "pursuing", "watch"]}
    }
    if theme_filter:
        query["$or"] = [
            {"themes_detected": theme_filter},
            {"themes": {"$regex": theme_filter.replace("_", " "), "$options": "i"}},
        ]
    if grant_type_filter:
        query["grant_type"] = grant_type_filter
    if status_filter:
        query["status"] = status_filter
    if min_funding is not None or max_funding is not None:
        fq: Dict[str, Any] = {}
        if min_funding is not None:
            fq["$gte"] = min_funding
        if max_funding is not None:
            fq["$lte"] = max_funding
        query["max_funding_usd"] = fq
    if search:
        sq = {"$regex": search, "$options": "i"}
        query["$or"] = [
            {"title": sq}, {"grant_name": sq},
            {"funder": sq}, {"geography": sq}, {"eligibility": sq},
        ]
    grants = list(
        db["grants_scored"].find(query, sort=[("weighted_total", -1)]).limit(500)
    )
    for g in grants:
        g["_id"] = str(g["_id"])
        # Normalise title: prefer grant_name (from new scraper) over title
        if not g.get("grant_name") and g.get("title"):
            g["grant_name"] = g["title"]
    return grants


def get_tracker_stats() -> Dict:
    """Summary stats for the Grant Tracker section."""
    db = _db()
    total     = db["grants_scored"].count_documents({})
    in_triage = db["grants_scored"].count_documents({"status": "triage"})
    pursuing  = db["grants_scored"].count_documents(
        {"status": {"$in": ["pursue", "pursuing"]}}
    )
    watching  = db["grants_scored"].count_documents({"status": "watch"})
    auto_pass = db["grants_scored"].count_documents(
        {"status": {"$in": ["auto_pass", "passed"]}}
    )
    # Capital targeted = sum max_funding_usd for pursue+watch+triage
    pipeline = db["grants_scored"].find(
        {"status": {"$in": ["triage", "pursue", "pursuing", "watch"]}},
        {"max_funding_usd": 1, "max_funding": 1},
    )
    capital = sum(
        float(g.get("max_funding_usd") or g.get("max_funding") or 0)
        for g in pipeline
    )
    return {
        "total": total,
        "in_triage": in_triage,
        "pursuing": pursuing,
        "watching": watching,
        "auto_pass": auto_pass,
        "capital_targeted": capital,
    }


def get_quick_stats() -> Dict:
    db = _db()
    total = db["grants_scored"].count_documents({})
    triage = db["grants_scored"].count_documents({"status": "triage"})
    pursuing = db["grants_scored"].count_documents({"status": {"$in": ["pursue", "pursuing"]}})
    last_scout = db["scout_runs"].find_one(sort=[("run_at", -1)])
    return {
        "total_grants": total,
        "in_triage": triage,
        "pursuing": pursuing,
        "last_scout": last_scout["run_at"] if last_scout else "Never",
    }


def get_dashboard_stats() -> Dict:
    db = _db()
    total = db["grants_scored"].count_documents({})
    triage = db["grants_scored"].count_documents({"status": "triage"})
    pursuing = db["grants_scored"].count_documents({"status": {"$in": ["pursue", "pursuing"]}})
    drafting = db["grants_pipeline"].count_documents({"status": "drafting"})
    complete = db["grants_pipeline"].count_documents({"status": "draft_complete"})

    # Total capital targeted (sum of max_funding for pursuing grants)
    pipeline_cur = db["grants_scored"].find({"status": {"$in": ["pursue", "pursuing", "drafting"]}})
    capital = sum(g.get("max_funding") or 0 for g in pipeline_cur)

    return {
        "total_discovered": total,
        "in_triage": triage,
        "pursuing": pursuing,
        "drafting": drafting,
        "draft_complete": complete,
        "capital_targeted": capital,
    }


def get_pipeline_funnel() -> List[Dict]:
    db = _db()
    stages = [
        ("Discovered", {}),
        ("In Triage", {"status": "triage"}),
        ("Pursuing", {"status": {"$in": ["pursue", "pursuing"]}}),
        ("Drafting", {"status": "drafting"}),
        ("Draft Complete", {"status": "draft_complete"}),
    ]
    scored = db["grants_scored"]
    pipeline_col = db["grants_pipeline"]
    result = []
    for label, query in stages:
        if label in ("Drafting", "Draft Complete"):
            count = pipeline_col.count_documents({"status": query.get("status", {})})
        else:
            count = scored.count_documents(query)
        result.append({"stage": label, "count": count})
    return result


def get_grants_by_theme() -> Dict[str, int]:
    db = _db()
    pipeline = [
        {"$unwind": "$themes_detected"},
        {"$group": {"_id": "$themes_detected", "count": {"$sum": 1}}},
        {"$sort": {"count": -1}},
    ]
    return {r["_id"]: r["count"] for r in db["grants_scored"].aggregate(pipeline)}


def get_score_distribution() -> List[float]:
    db = _db()
    return [g["weighted_total"] for g in db["grants_scored"].find({}, {"weighted_total": 1})]


def get_recent_activity(limit: int = 8) -> List[Dict]:
    db = _db()
    return list(db["audit_logs"].find({}, sort=[("created_at", -1)]).limit(limit))


def get_top_grants(limit: int = 5) -> List[Dict]:
    db = _db()
    grants = list(
        db["grants_scored"].find(
            {"recommended_action": {"$in": ["pursue", "watch"]}},
            sort=[("weighted_total", -1)],
        ).limit(limit)
    )
    for g in grants:
        g["_id"] = str(g["_id"])
    return grants


# ── Manual Grant Entry ─────────────────────────────────────────────────────────

def save_manual_grant(
    url: str,
    title_override: str = "",
    funder_override: str = "",
    notes: str = "",
    jina_key: str = "",
) -> tuple:
    """Fetch a grant URL via Jina, detect themes, and save to grants_raw.

    Returns (success: bool, message: str).
    """
    import hashlib
    import httpx
    from urllib.parse import urlparse

    url = url.strip()
    if not url.startswith("http"):
        return False, "URL must start with http:// or https://"

    url_hash = hashlib.md5(url.lower().encode()).hexdigest()

    existing = _db()["grants_raw"].find_one({"url_hash": url_hash})
    if existing:
        return False, f"Already in database (added {existing.get('scraped_at','previously')[:10]})."

    # Fetch via Jina first, fall back to plain HTTP
    raw_content = ""
    jina_url = f"https://r.jina.ai/{url}"
    headers = {"X-Return-Format": "markdown", "X-With-Links-Summary": "false"}
    if jina_key:
        headers["Authorization"] = f"Bearer {jina_key}"
    try:
        r = httpx.get(jina_url, headers=headers, timeout=30.0, follow_redirects=True)
        r.raise_for_status()
        raw_content = r.text.strip()[:80_000]
    except Exception as e:
        try:
            r2 = httpx.get(url, headers={"User-Agent": "Mozilla/5.0"}, timeout=20.0, follow_redirects=True)
            r2.raise_for_status()
            raw_content = r2.text[:60_000]
        except Exception:
            return False, f"Could not fetch URL: {e}"

    if len(raw_content) < 100:
        return False, "Page returned too little content. Check the URL."

    # Theme detection
    text_lower = (raw_content + " " + title_override).lower()
    themes = []
    if any(k in text_lower for k in ["climate", "carbon", "net zero", "decarboni", "emission", "cdr", "mrv", "cleantech", "renewable"]):
        themes.append("climatetech")
    if any(k in text_lower for k in ["agri", "soil", "farm", "crop", "food", "land use", "regenerative"]):
        themes.append("agritech")
    if any(k in text_lower for k in ["artificial intelligence", "machine learning", "ai for", "deep learning", "nlp"]):
        themes.append("ai_for_sciences")
    if any(k in text_lower for k in ["earth science", "remote sensing", "satellite", "geology", "geospatial", "subsurface"]):
        themes.append("applied_earth_sciences")
    if any(k in text_lower for k in ["social impact", "community", "rural", "livelihood", "inclusive", "women", "development"]):
        themes.append("social_impact")
    if not themes:
        themes = ["climatetech"]

    # Auto-extract title from content if not provided
    title = title_override.strip()
    if not title:
        import re as _re
        # Try <title> tag first (for raw HTML fallback)
        m = _re.search(r"<title[^>]*>([^<]+)</title>", raw_content, _re.IGNORECASE)
        if m:
            title = m.group(1).strip()[:120]
        else:
            # For markdown (Jina output), use the first non-empty heading or line
            for line in raw_content.split("\n"):
                line = line.lstrip("#").strip()
                if len(line) > 10:
                    title = line[:120]
                    break
        if not title:
            title = url

    # Auto-extract funder from domain if not provided
    funder = funder_override.strip()
    if not funder:
        try:
            domain = urlparse(url).netloc.replace("www.", "")
            funder = domain.split(".")[0].upper()
        except Exception:
            funder = "Unknown"

    doc = {
        "title": title,
        "url": url,
        "url_hash": url_hash,
        "funder": funder,
        "raw_content": raw_content,
        "themes_detected": themes,
        "source": "manual",
        "deadline": None,
        "max_funding": None,
        "currency": "USD",
        "eligibility_raw": "",
        "processed": False,
        "notes": notes,
        "scraped_at": datetime.now(timezone.utc).isoformat(),
    }
    _db()["grants_raw"].insert_one(doc)
    return True, (
        f"Saved **{title[:70]}** · {len(raw_content):,} chars fetched · "
        f"themes: {', '.join(themes)}"
    )


# ── Grants Pipeline ────────────────────────────────────────────────────────────

def get_all_pipeline_grants(
    search: str = "",
    theme_filter: str = "",
    status_filter: str = "",
    grant_type_filter: str = "",
    min_funding: Optional[int] = None,
    max_funding: Optional[int] = None,
) -> List[Dict]:
    db = _db()
    query: Dict[str, Any] = {}
    if theme_filter:
        query["themes_detected"] = theme_filter
    if status_filter:
        query["status"] = status_filter
    if grant_type_filter:
        query["grant_type"] = grant_type_filter
    if min_funding is not None or max_funding is not None:
        funding_q: Dict[str, Any] = {}
        if min_funding is not None:
            funding_q["$gte"] = min_funding
        if max_funding is not None:
            funding_q["$lte"] = max_funding
        query["max_funding"] = funding_q
    if search:
        query["$or"] = [
            {"title":       {"$regex": search, "$options": "i"}},
            {"funder":      {"$regex": search, "$options": "i"}},
            {"geography":   {"$regex": search, "$options": "i"}},
            {"eligibility": {"$regex": search, "$options": "i"}},
        ]
    grants = list(db["grants_scored"].find(query, sort=[("weighted_total", -1)]).limit(200))
    for g in grants:
        g["_id"] = str(g["_id"])
    return grants


def update_grant_status(grant_id: str, status: str):
    from bson import ObjectId
    _db()["grants_scored"].update_one({"_id": ObjectId(grant_id)}, {"$set": {"status": status}})


REPORT_REASONS = [
    "Not relevant to AltCarbon",
    "News article, not a grant",
    "Wrong / incomplete info scraped",
    "Grant is already closed / expired",
    "Duplicate grant",
    "Other",
]


def report_grant(grant_id: str, reason: str, note: str = ""):
    """Flag a grant as reported — removes it from active queues."""
    from bson import ObjectId
    db = _db()
    db["grants_scored"].update_one(
        {"_id": ObjectId(grant_id)},
        {
            "$set": {"status": "reported"},
            "$push": {
                "reports": {
                    "reason": reason,
                    "note": note,
                    "reported_at": datetime.now(timezone.utc).isoformat(),
                }
            },
        },
    )


# ── Triage Queue ───────────────────────────────────────────────────────────────

def get_triage_queue(
    search: str = "",
    theme_filter: str = "",
    grant_type_filter: str = "",
    min_score: float = 0.0,
    min_funding: Optional[int] = None,
    max_funding: Optional[int] = None,
) -> List[Dict]:
    db = _db()
    query: Dict[str, Any] = {"status": "triage"}
    if theme_filter:
        query["themes_detected"] = theme_filter
    if grant_type_filter:
        query["grant_type"] = grant_type_filter
    if min_score > 0:
        query["weighted_total"] = {"$gte": min_score}
    if min_funding is not None or max_funding is not None:
        funding_q: Dict[str, Any] = {}
        if min_funding is not None:
            funding_q["$gte"] = min_funding
        if max_funding is not None:
            funding_q["$lte"] = max_funding
        query["max_funding"] = funding_q
    if search:
        query["$or"] = [
            {"title":       {"$regex": search, "$options": "i"}},
            {"funder":      {"$regex": search, "$options": "i"}},
            {"geography":   {"$regex": search, "$options": "i"}},
            {"eligibility": {"$regex": search, "$options": "i"}},
        ]
    grants = list(db["grants_scored"].find(query, sort=[("weighted_total", -1)]))
    for g in grants:
        g["_id"] = str(g["_id"])
    return grants


# ── Drafter ────────────────────────────────────────────────────────────────────

def get_active_drafts() -> List[Dict]:
    db = _db()
    pipelines = list(
        db["grants_pipeline"].find(
            {"status": {"$in": ["drafting", "draft_complete"]}},
            sort=[("started_at", -1)],
        )
    )
    result = []
    for p in pipelines:
        p["_id"] = str(p["_id"])
        grant = db["grants_scored"].find_one({"_id": __import__("bson").ObjectId(p["grant_id"])}) if p.get("grant_id") else {}
        p["grant_title"] = (grant or {}).get("title", "Unknown Grant")
        p["grant_funder"] = (grant or {}).get("funder", "")
        # Get latest draft
        draft = db["grant_drafts"].find_one(
            {"pipeline_id": p["_id"]},
            sort=[("version", -1)],
        )
        if draft:
            draft["_id"] = str(draft["_id"])
        p["latest_draft"] = draft
        result.append(p)
    return result


def get_thread_interrupt(thread_id: str) -> Optional[Dict]:
    """Get the current pending section interrupt for a thread."""
    import json
    from bson import ObjectId
    doc = _db()["graph_checkpoints"].find_one(
        {"thread_id": thread_id},
        sort=[("checkpoint_id", -1)],
    )
    if not doc:
        return None
    checkpoint = json.loads(doc.get("checkpoint", "{}"))
    channel_values = checkpoint.get("channel_values", {})
    return channel_values.get("pending_interrupt")


# ── Knowledge Health ───────────────────────────────────────────────────────────

def knowledge_base_health() -> Dict:
    db = _db()
    total = db["knowledge_chunks"].count_documents({})
    notion = db["knowledge_chunks"].count_documents({"source": "notion"})
    drive = db["knowledge_chunks"].count_documents({"source": "drive"})
    past_grants = db["knowledge_chunks"].count_documents({"doc_type": "past_grant_application"})

    by_type_pipeline = [{"$group": {"_id": "$doc_type", "count": {"$sum": 1}}}]
    by_type = {r["_id"]: r["count"] for r in db["knowledge_chunks"].aggregate(by_type_pipeline)}

    by_theme_pipeline = [
        {"$unwind": "$themes"},
        {"$group": {"_id": "$themes", "count": {"$sum": 1}}},
    ]
    by_theme = {r["_id"]: r["count"] for r in db["knowledge_chunks"].aggregate(by_theme_pipeline)}

    last_sync = db["knowledge_sync_logs"].find_one(sort=[("synced_at", -1)])
    status = "healthy" if total >= 200 else ("thin" if total >= 50 else "critical")

    return {
        "total_chunks": total,
        "notion_chunks": notion,
        "drive_chunks": drive,
        "past_grant_application_chunks": past_grants,
        "by_type": by_type,
        "by_theme": by_theme,
        "last_synced": last_sync["synced_at"] if last_sync else None,
        "status": status,
    }


def get_sync_logs(limit: int = 5) -> List[Dict]:
    db = _db()
    logs = list(db["knowledge_sync_logs"].find({}, sort=[("synced_at", -1)]).limit(limit))
    for l in logs:
        l["_id"] = str(l["_id"])
    return logs


# ── Agent Config ───────────────────────────────────────────────────────────────

def get_agent_config(agent: str = "") -> Dict:
    db = _db()
    if agent:
        return db["agent_config"].find_one({"agent": agent}) or {}
    return {r["agent"]: r for r in db["agent_config"].find({})}


def save_agent_config(agent: str, config: Dict):
    config["updated_at"] = datetime.now(timezone.utc).isoformat()
    _db()["agent_config"].update_one({"agent": agent}, {"$set": config}, upsert=True)


# ── Export ─────────────────────────────────────────────────────────────────────

def export_grants_csv() -> bytes:
    import csv, io
    grants = get_all_pipeline_grants()
    output = io.StringIO()
    if not grants:
        return b""
    fieldnames = [
        "title", "funder", "grant_type", "geography",
        "amount", "max_funding", "currency",
        "deadline", "eligibility",
        "weighted_total", "status", "recommended_action",
        "rationale", "reasoning",
        "url", "application_url",
        "themes_detected",
    ]
    writer = csv.DictWriter(output, fieldnames=fieldnames, extrasaction="ignore")
    writer.writeheader()
    for g in grants:
        row = {k: g.get(k, "") for k in fieldnames}
        # Flatten lists
        if isinstance(row.get("themes_detected"), list):
            row["themes_detected"] = " | ".join(row["themes_detected"])
        writer.writerow(row)
    return output.getvalue().encode()

File: Grant_engine-2/frontend/src/lib/queries.ts
/**
 * TypeScript port of app/db/queries.py — all MongoDB read functions for the UI.
 * Import only in Server Components, Server Actions, and API routes.
 */
import { getDb } from "./mongodb";

// ── Types ──────────────────────────────────────────────────────────────────────

export interface Grant {
  _id: string;
  grant_name?: string;
  title?: string;
  funder?: string;
  status: string;
  weighted_total?: number;
  deadline_urgent?: boolean;
  deadline?: string;
  geography?: string;
  eligibility?: string;
  max_funding_usd?: number;
  max_funding?: number;
  themes_detected?: string[];
  recommended_action?: string;
  rationale?: string;
  scores?: Record<string, number>;
  human_override?: boolean;
  override_reason?: string;
  override_at?: string;
  days_to_deadline?: number;
  thread_id?: string;
  grant_type?: string;
  url?: string;
  application_url?: string;
  scored_at?: string;
  scraped_at?: string;
}

export interface PipelineRecord {
  _id: string;
  grant_id: string;
  thread_id: string;
  status: string;
  started_at?: string;
  draft_started_at?: string;
  current_draft_version?: number;
  final_draft_url?: string | null;
  grant_title?: string;
  grant_funder?: string;
  latest_draft?: DraftDoc | null;
}

export interface DraftSection {
  content: string;
  approved?: boolean;
  revision_count?: number;
  word_count?: number;
}

export interface DraftDoc {
  _id: string;
  pipeline_id: string;
  version: number;
  sections: Record<string, DraftSection>;
  created_at?: string;
}

export interface KnowledgeStatus {
  total_chunks: number;
  notion_chunks: number;
  drive_chunks: number;
  past_grant_chunks: number;
  by_type: Record<string, number>;
  by_theme: Record<string, number>;
  last_synced: string | null;
  status: "healthy" | "thin" | "critical";
}

export interface AgentConfig {
  agent: string;
  [key: string]: unknown;
}

// ── Helpers ────────────────────────────────────────────────────────────────────

function serializeId(doc: Record<string, unknown>): Record<string, unknown> {
  return { ...doc, _id: String(doc._id) };
}

// ── Dashboard ──────────────────────────────────────────────────────────────────

export async function getDashboardStats() {
  const db = await getDb();
  const [total, triage, pursuing, watching, onHold, drafting, complete, urgentCount] =
    await Promise.all([
      db.collection("grants_scored").countDocuments({}),
      db.collection("grants_scored").countDocuments({ status: "triage" }),
      db.collection("grants_scored").countDocuments({ status: { $in: ["pursue", "pursuing"] } }),
      db.collection("grants_scored").countDocuments({ status: "watch" }),
      db.collection("grants_scored").countDocuments({ status: "hold" }),
      db.collection("grants_pipeline").countDocuments({ status: "drafting" }),
      db.collection("grants_pipeline").countDocuments({ status: "draft_complete" }),
      db.collection("grants_scored").countDocuments({
        deadline_urgent: true,
        status: { $in: ["triage", "pursue", "watch"] },
      }),
    ]);

  const warnings: string[] = [];
  if (total > 0 && triage === 0) {
    warnings.push(
      "Shortlist is empty — run a new scout to discover fresh opportunities."
    );
  }
  if (urgentCount > 0) {
    warnings.push(
      `${urgentCount} grant(s) with urgent deadlines (≤30 days) in your active queue — review now.`
    );
  }
  if (onHold > 0) {
    warnings.push(`${onHold} grant(s) on HOLD due to unresolved currency — manual review needed.`);
  }

  return {
    total_discovered: total,
    in_triage: triage,
    pursuing,
    watching,
    on_hold: onHold,
    deadline_urgent_count: urgentCount,
    drafting,
    draft_complete: complete,
    warnings,
  };
}

export async function getGrantsActivity(
  days = 30
): Promise<{ date: string; count: number }[]> {
  const db = await getDb();
  const since = new Date(Date.now() - days * 24 * 60 * 60 * 1000).toISOString();

  const result = await db
    .collection("grants_raw")
    .aggregate([
      { $match: { scraped_at: { $gte: since } } },
      {
        $group: {
          _id: { $substr: ["$scraped_at", 0, 10] },
          count: { $sum: 1 },
        },
      },
      { $sort: { _id: 1 } },
    ])
    .toArray();

  return result.map((r) => ({ date: r._id as string, count: r.count as number }));
}

// ── Pipeline Kanban ────────────────────────────────────────────────────────────

export async function getPipelineGrants(): Promise<Record<string, Grant[]>> {
  const db = await getDb();
  // Fetch ALL scored grants — table view shows auto-passed too
  const docs = await db
    .collection("grants_scored")
    .find({})
    .sort({ weighted_total: -1 })
    .limit(1000)
    .toArray();

  const grouped: Record<string, Grant[]> = {
    shortlisted: [],
    pursue: [],
    watch: [],
    drafting: [],
    submitted: [],
    passed: [],
  };

  for (const doc of docs) {
    const g = serializeId(doc as Record<string, unknown>) as unknown as Grant;
    if (!g.grant_name && g.title) g.grant_name = g.title;

    if (g.status === "triage") grouped.shortlisted.push(g);
    else if (g.status === "pursue" || g.status === "pursuing") grouped.pursue.push(g);
    else if (g.status === "watch") grouped.watch.push(g);
    else if (g.status === "drafting") grouped.drafting.push(g);
    else if (g.status === "draft_complete" || g.status === "submitted" || g.status === "won")
      grouped.submitted.push(g);
    else if (g.status === "passed" || g.status === "auto_pass" || g.status === "human_passed" || g.status === "reported")
      grouped.passed.push(g);
  }

  return grouped;
}

// ── Triage Queue ───────────────────────────────────────────────────────────────

export async function getTriageQueue(): Promise<Grant[]> {
  const db = await getDb();
  const docs = await db
    .collection("grants_scored")
    .find({ status: "triage" })
    .sort({ weighted_total: -1 })
    .toArray();

  return docs.map((doc) => {
    const g = serializeId(doc as Record<string, unknown>) as unknown as Grant;
    if (!g.grant_name && g.title) g.grant_name = g.title;
    return g;
  });
}

// ── Drafter ────────────────────────────────────────────────────────────────────

export async function getDraftGrants(): Promise<PipelineRecord[]> {
  const db = await getDb();
  const pipelines = await db
    .collection("grants_pipeline")
    .find({ status: { $in: ["drafting", "draft_complete"] } })
    .sort({ started_at: -1 })
    .toArray();

  const result: PipelineRecord[] = [];

  for (const p of pipelines) {
    const pid = serializeId(p as Record<string, unknown>) as unknown as PipelineRecord;

    // Enrich with grant title/funder
    if (p.grant_id) {
      const { ObjectId } = await import("mongodb");
      const grant = await db
        .collection("grants_scored")
        .findOne({ _id: new ObjectId(p.grant_id as string) });
      if (grant) {
        pid.grant_title = (grant.grant_name as string) || (grant.title as string) || "Unknown Grant";
        pid.grant_funder = (grant.funder as string) || "";
      }
    }

    // Attach latest draft
    const draft = await db
      .collection("grant_drafts")
      .findOne({ pipeline_id: pid._id }, { sort: { version: -1 } });
    if (draft) {
      pid.latest_draft = serializeId(draft as Record<string, unknown>) as unknown as DraftDoc;
    }

    result.push(pid);
  }

  return result;
}

export async function getSections(pipelineId: string): Promise<Record<string, DraftSection>> {
  const db = await getDb();
  const draft = await db
    .collection("grant_drafts")
    .findOne({ pipeline_id: pipelineId }, { sort: { version: -1 } });
  if (!draft) return {};
  return (draft.sections as Record<string, DraftSection>) || {};
}

// ── Knowledge Health ───────────────────────────────────────────────────────────

export async function getKnowledgeStatus(): Promise<KnowledgeStatus> {
  const db = await getDb();
  const [total, notion, drive, pastGrants] = await Promise.all([
    db.collection("knowledge_chunks").countDocuments({}),
    db.collection("knowledge_chunks").countDocuments({ source: "notion" }),
    db.collection("knowledge_chunks").countDocuments({ source: "drive" }),
    db.collection("knowledge_chunks").countDocuments({ doc_type: "past_grant_application" }),
  ]);

  const byTypeAgg = await db
    .collection("knowledge_chunks")
    .aggregate([{ $group: { _id: "$doc_type", count: { $sum: 1 } } }])
    .toArray();
  const byType: Record<string, number> = {};
  for (const r of byTypeAgg) byType[r._id as string] = r.count as number;

  const byThemeAgg = await db
    .collection("knowledge_chunks")
    .aggregate([
      { $unwind: "$themes" },
      { $group: { _id: "$themes", count: { $sum: 1 } } },
    ])
    .toArray();
  const byTheme: Record<string, number> = {};
  for (const r of byThemeAgg) byTheme[r._id as string] = r.count as number;

  const lastSync = await db
    .collection("knowledge_sync_logs")
    .findOne({}, { sort: { synced_at: -1 } });

  const status: KnowledgeStatus["status"] =
    total >= 200 ? "healthy" : total >= 50 ? "thin" : "critical";

  return {
    total_chunks: total,
    notion_chunks: notion,
    drive_chunks: drive,
    past_grant_chunks: pastGrants,
    by_type: byType,
    by_theme: byTheme,
    last_synced: lastSync ? (lastSync.synced_at as string) : null,
    status,
  };
}

export async function getSyncLogs(limit = 5) {
  const db = await getDb();
  const docs = await db
    .collection("knowledge_sync_logs")
    .find({})
    .sort({ synced_at: -1 })
    .limit(limit)
    .toArray();
  return docs.map((d) => serializeId(d as Record<string, unknown>));
}

// ── Agent Config ───────────────────────────────────────────────────────────────

export async function getAgentConfig(agent?: string): Promise<AgentConfig | Record<string, AgentConfig>> {
  const db = await getDb();
  if (agent) {
    const doc = await db.collection("agent_config").findOne({ agent });
    if (!doc) return { agent };
    return serializeId(doc as Record<string, unknown>) as unknown as AgentConfig;
  }
  const docs = await db.collection("agent_config").find({}).toArray();
  const result: Record<string, AgentConfig> = {};
  for (const d of docs) {
    const cfg = serializeId(d as Record<string, unknown>) as unknown as AgentConfig;
    result[cfg.agent] = cfg;
  }
  return result;
}

export async function getGrantById(id: string): Promise<Grant | null> {
  const db = await getDb();
  const { ObjectId } = await import("mongodb");
  try {
    const doc = await db.collection("grants_scored").findOne({ _id: new ObjectId(id) });
    if (!doc) return null;
    const g = serializeId(doc as Record<string, unknown>) as unknown as Grant;
    if (!g.grant_name && g.title) g.grant_name = g.title;
    return g;
  } catch {
    return null;
  }
}

export async function saveAgentConfig(agent: string, config: Record<string, unknown>) {
  const db = await getDb();
  const update = { ...config, agent, updated_at: new Date().toISOString() };
  await db.collection("agent_config").updateOne({ agent }, { $set: update }, { upsert: true });
}
